### PR TITLE
Dynamically resize line number gutter width

### PR DESCRIPTION
**Pull request by dgkf**
Addressing #2583, this PR implements the second proposed option.

### Behavior

- The line number gutter will resize based on the number of digits in the line count for the Document
- The line number gutter will have a minimum width of 2 characters. This is so that configurations using `line-number: "relative"` aren't adversely affected by this change. In the future, the min/max width could perhaps be configurable and this hard-coded minimum could be removed.
- Enabled the `GutterType::Spacer` option. (I think the default `gutter` config should probably become `["diagnostics", "spacer", "line-number"]` now numbers will butt right up against diagnostics past line 100)

_behaviors that have been modified after review comments_

- ~~If the number of digits on screen exceeds 5 digits, only the last 4 digits are displayed (preceded by `…`)~~ no longer truncates (as suggested in https://github.com/helix-editor/helix/pull/3469#discussion_r948628101)
- ~~Line number gutter width based on _view_ bottom, not _document_ bottom.~~ Fixed

# Screenshots

image
> A very short window to highlight shrinking width (`gutters: ["diagnostics", "spacer", "line-number"]`)

image
> With default config, diagnostic symbols look a bit cramped

### CODE REVIEW

File: helix-term/src/ui/editor.rs
Comment (dgkf) [inline helix-term/src/ui/editor.rs:643]: `width` is no longer passed to `gutter_type.row_styler` (formerly `constructor`), though I could see the argument for calculating `width` first and providing it to the `row_styler`.

File: helix-view/src/gutter.rs
Comment (dgkf) [inline helix-view/src/gutter.rs:156]: It would be great to have `Document` available here to base this line number width on `min(<last_document_line>, <last_visible_line>)`, but I quickly ended up with cyclic data flows. Total gutter width is needed for `view.inner_area` is needed for last line is needed for total gutter width.

**Comment by archseer** [comment]
> * The line number gutter will resize based on the number of digits for the line _at the bottom of the view*_

Bottom of the document is what we want though, otherwise the document will jitter sideways as you scroll.

File: helix-view/src/gutter.rs
Comment (archseer) [inline helix-view/src/gutter.rs:96]: Wouldn't it be sufficient to call this with `text.len_lines()`, without having to refactor?

File: helix-view/src/gutter.rs
Comment (archseer) [inline helix-view/src/gutter.rs:12]: Fun snippet :) I'd fully qualify `std::iter::successors` rather than importing since it's only used in one location

Comment (archseer) [reply helix-view/src/gutter.rs:12]: I'd also add a note that this can be simplified to `n.checked_log10().unwrap_or(0) + 1` if `int_log` gets standardized in stdlib

File: helix-view/src/gutter.rs
Comment (archseer) [inline helix-view/src/gutter.rs:144]: I don't think we need truncation, line numbers should always be fully visible. I can't think of any other editor truncating, past 8 digits the file will be enormously large anyway.

File: helix-view/src/view.rs
Comment (archseer) [inline helix-view/src/view.rs:118]: This resolves the TODO on the line above

File: helix-view/src/gutter.rs
Comment (dgkf) [reply helix-view/src/gutter.rs:12]: Wow! You all work fast. Stepped away to pet my cat and had a full review when I got back on a PR in draft. Kudos to you!

I'll qualify the use and add a note that this can be done less opaquely once `int_log` is standardized.

File: helix-view/src/gutter.rs
Comment (dgkf) [reply helix-view/src/gutter.rs:144]: Alright, that's easy enough. The design was taken from the discussion in #2583 - happy to change it if it's preferred that it never truncates.

File: helix-view/src/gutter.rs
Comment (dgkf) [reply helix-view/src/gutter.rs:96]: Sure, that works. I was basing it off of the first suggestion here ([2438](https://github.com/helix-editor/helix/issues/2583#issuecomment-1139625428)), but I can do the latter suggestion instead. 

> make the width dynamic based on the line numbers within view or the maximum line number of the buffer

Comment (dgkf) [reply helix-view/src/gutter.rs:94]: I'm having trouble making this fix. 

It's important that the width here is the same as how it's calculated for `GutterType.width` used to calculate `gutter_offset`. If these provide different widths, the gutter and numbers don't stay aligned.

But I couldn't find an elegant way to provide the `Document` to `GutterType::LineNumber.width` in a way that doesn't require some major changes. But I'm still quite new to the code base, so perhaps there's some obvious solution I'm missing.

Comment (dgkf) [reply helix-view/src/gutter.rs]: Revisited this and it ended up being less problematic than I first thought. I needed to pass the `Document` to `View.inner_area`. This seemed fine in all but one place, where I think the total `View` area is a fine substitution. I'll highlight it in a separate thread.

File: helix-view/src/tree.rs
Comment (dgkf) [inline helix-view/src/tree.rs:485]: This is the only place where the Document isn't easily available to calculate the inner area (now that the gutter width is based on the Document contents). A simple solution is to use the entire view area instead. 

I think this would affect things like moving to a split when the screen location that you would move to would be in the gutter of a neighboring split. Previously, if you were to try to find a neighboring split at a gutter, you would get the window to the left, now you would get the window that contains the gutter.

File: helix-term/src/ui/editor.rs
Comment (the-mikedavis) [inline helix-term/src/ui/editor.rs:740]: I don't think this is a general rule in the codebase but for me "row_styler" comes across as https://wiki.c2.com/?DontCreateVerbClasses

I prefer the old name of "constructor" but maybe there's a better word for it: "style" or something similar

File: helix-view/src/gutter.rs
Comment (the-mikedavis) [inline helix-view/src/gutter.rs:10]: Looks like this entered the FCP the other day https://github.com/rust-lang/rust/issues/70887#issuecomment-1289028216

Could you link the issue in this comment too to raise the chances of it being refactored later?

File: helix-term/src/ui/editor.rs
Comment (dgkf) [reply helix-term/src/ui/editor.rs]: Thanks for the tip! I agree, this name never felt quite right to me. 

When entering into the code base, I found "constructor" to be a bit of a misnomer. I went with the suggested "style", which I quite like.

**Comment by the-mikedavis** [comment]
I've been testing this for a bit and I think you're right about the diagnostics being too cramped when there aren't many lines in a file. I think changing the default gutters to include a spacer between the diagnostics and line-numbers elements looks much better 👍

**Comment by dgkf** [comment]
Thanks @the-mikedavis - I forgot that was even non-standard. I've been using this with the spacer for so long now. I'll update with the change to the default.

**Comment by dgkf** [comment]
@the-mikedavis, the gutter `spacer` has been added as default (separating the diagnostics from the line numbers). Let me know if you want anything else to change before it's in an acceptable place.

File: helix-term/src/commands.rs
Comment (archseer) [inline helix-term/src/commands.rs:873]: In almost every circumstance we're only interested in the height which is unaffected by gutters. Let's add a view.height() method that calculates it without having to pass in a doc. Additionally the method should return a usize since we're casting to usize everywhere

File: helix-view/src/view.rs
Comment (archseer) [inline helix-view/src/view.rs:143]: This for loop could be replaced with an iter.map().sum()

Comment (dgkf) [reply helix-view/src/view.rs]: swapped for iterator equivalent

File: helix-term/src/commands.rs
Comment (dgkf) [reply helix-term/src/commands.rs]: Sounds good - I added a `view.inner_height()` which is just slightly different than a `view.height()` method because it subtracts the statusbar height allowing it to be dropped in for all the places where `inner_area().height` was being used previously.

File: helix-term/src/commands.rs
Comment (the-mikedavis) [inline helix-term/src/commands.rs:1421]: These can all be reset to `let view = view!(cx.editor)`

Comment (dgkf) [reply helix-term/src/commands.rs]: Thanks - good catch. Apologies for the sloppiness here. It's been a while since I kicked off the work and the details escape me now.

**Review by the-mikedavis — APPROVED** [review]

**Comment by dgkf** [comment]
@archseer - Any chance I could grab your attention on this one? 

Along those lines, I can't tell if I'm being annoying or helpful when I ping you all or re-request review. I know you all have an absolutely enormous PR backlog (IMO a reflection on the community's enthusiasm and not the pace of the maintainers :tada:), and I can't tell what is the best way to grab your attention.

After this PR, would it be helpful to submit a "PR Etiquette" section to the CONTRIBUTING guide to make this relationship less ambiguous? I can draft something up, but you all would probably need to weigh in on how you want things to flow.

**Comment by archseer** [comment]
I think in this case a ping is fine because I was going to merge this once the last change was addressed :)

**Review by archseer — APPROVED** [review]
Thanks for working on this!

**Pull request merged**
